src/server/io_utils.h

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "base/io_buf.h"
+#include "io/io.h"
+#include "server/common.h"
+#include "util/fibers/event_count.h"
+
+namespace dfly {
+
+// Base for constructing buffered byte streams with backpressure
+// for single producer and consumer on the same thread.
+//
+// Use it as a io::Sink to write data from a producer fiber,
+// and ConsumeIntoSink to extract this data in a consumer fiber.
+// Use NotifyWritten to request the consumer to be woken up.
+//
+// Uses two base::IoBuf internally that are swapped in turns.
+class BufferedStreamerBase : public io::Sink {
+ protected:
+  // Initialize with global cancellation and optional stall conditions.
+  // Pointer to external buffer is set separatly because of sub-intialization order.
+  BufferedStreamerBase(const Cancellation* cll, unsigned max_buffered_cnt = 5,
+                       unsigned max_buffered_mem = 512)
+      : cll_{cll}, max_buffered_cnt_{max_buffered_cnt}, max_buffered_mem_{max_buffered_mem} {
+  }
+
+  // Write some data into the internal buffer.
+  // Consumer needs to be woken up manually with NotifyWritten to avoid waking it up for small
+  // writes.
+  io::Result<size_t> WriteSome(const iovec* vec, uint32_t len) override;
```

Comment:
Because notify issues a wakeup which I don't always want.

Now that the writer is unbuffered, we don't want a wakeup after writing every small part. In the future we even might now want to wake after ever entry